### DISCUSSION TITLE
Incompatible type for `get_or_create` vs. `update`?

### DISCUSSION THREAD

**Discussion by eddyg** [piccolo-orm/piccolo#792] [Q&A] [answered]
I'm not a Python typing expert, but I'm wondering how to resolve this.

Given:

```python
    class Todo(Table):
        todo_id = Integer(primary_key=True)
        item = Text()
        assignee = Text(null=True)
        last_updated = Timestamp()
```

and then:

```python
       values: dict[Column, Any] = {
          Todo.item: text,
          Todo.assignee: user_id,
          Todo.last_updated: datetime.now(timezone.utc),
       }

       if response := await Todo.objects().get_or_create(Todo.todo_id == todo_id, values):
           if not response._was_created:
               await Todo.update(values).where(Todo.todo_id == todo_id)
```

*mypy* complains that:

```
error: Argument 1 to "update" of "Table" has incompatible type "Dict[Column, Any]"; expected "Dict[Union[Column, str], Any]"  [arg-type]
```

FWIW, _without_ the explicit `dict[Column, Any]` type spec on `values`,  then [Pyright](https://github.com/microsoft/pyright) (the real-time static type checker used in Pylance in VS Code) also complains about the type of `get_or_create`'s "default" parameter, but this may be a Pyright issue:

```
Argument of type "dict[Text | Timestamp, Unknown]" cannot be assigned to parameter "defaults" of type "Dict[Column, Any]" in function "get_or_create"
  "dict[Text | Timestamp, Unknown]" is incompatible with "Dict[Column, Any]"
    TypeVar "_KT@dict" is invariant
      Type "Text | Timestamp" cannot be assigned to type "Column"
        "Text" is incompatible with "Column"
```

Any suggestions on at least getting the *mypy* error to go away?

**Comment by dantownsend** [comment, marked as answer]
Yeah, this is a bit of a pain. I'd just ignore it for now using `# type: ignore`.

`dict` is invariant so is very unforgiving. I'll change this line:

https://github.com/piccolo-orm/piccolo/blob/56669fab0cc6626ccc7be707b094a022b9a0ba36/piccolo/table.py#L1115

To:

```python
values: t.Union[t.Mapping[Column, t.Any], t.Mapping[str, t.Any], None] = None,
```

**Reply by eddyg** [reply]
Thanks again for your help!

For now I've ignored the typing issue, but appreciate that you'll update the type in a future release. (FWIW, I know it's a pain, but I really like that Piccolo is typed. It definitely helps avoid coding errors!)
